### PR TITLE
feat: add data request incentives

### LINKED ISSUE #77: ✨ Implement Post Request Incentives and Refunds
## ✨ Feature

Implement an incentive mechanism for posting requests, along with a refund system to ensure fairness and efficiency.

## Motivation

This feature incentivizes users to post requests by attaching a reward. It also ensures refunds are provided when requests are not fulfilled or conditions are not met.

## Implementation

- Add logic to attach incentives when a request is posted.
- Implement a refund mechanism that returns to the request creator the amount not used (derived from gas used fields)
- Update or extend interfaces as needed.
- Emit relevant events for incentives and refunds.

### LINKED ISSUE #78: ✨ Implement Post Result Incentive
## ✨ Feature

Implement an incentive mechanism to reward users for posting valid results.

## Motivation

This feature ensures that users are incentivized to provide accurate and timely results for posted requests. It directly affects the result submission functions.

## Implementation

- Add logic to reward users when valid results are posted.
- Verify conditions for result validity before granting incentives.
- Emit relevant events when incentives are distributed.
- Update or extend interfaces if necessary.

### LINKED ISSUE #79: ✨ Implement Post Batch Incentive
## ✨ Feature

Implement an incentive mechanism to reward users for posting valid batches.

## Motivation

This feature encourages users to submit batches. It affects the batch submission function.

## Implementation

- Add logic to reward users for posting valid batches.
- Ensure batches meet defined conditions before granting incentives.
- Emit relevant events when batch incentives are distributed.
- Update or extend interfaces if required.

### LINKED ISSUE #80: ✨ Implement Update Fee(s) Mechanism
## ✨ Feature

Implement a mechanism to allow updating fees dynamically for posting requests, results, or batches.

## Motivation

This feature ensures flexibility in adjusting fees based on network conditions, usage patterns, or protocol changes.

## Implementation

- Add logic to update fees for:
  - Posting requests.
  - Posting results.
  - Posting batches.
- ~~Ensure only authorized roles can update fees.~~
- Emit relevant events when fees are updated.
- Update or extend interfaces as needed.

### PR DESCRIPTION
**Pull request by mariocao**
## Motivation

This PR implements fee management functionality in the SEDA Core contract, allowing for:
- Request fees with gas-based refunds (derived from data results)
- Result fees for result solvers
- Batch fees for batch solver
- Support of fee increases for pending requests

## Explanation of Changes

The changes introduce several key components:
- Added fee distribution logic in `postResult` function
- Implemented gas-based refund mechanism for request fees
- Added `increaseFees` function to allow updating fees for pending requests
- Added batch sender tracking for proper fee distribution
- Introduced `RequestDetails` struct to efficiently store fee-related data
- Added relevant events for fee distribution tracking
- Updated interfaces to support fee functionality

## Testing

The changes are covered by comprehensive tests:
  - Fee distribution logic
  - Gas-based refund calculations
  - Fee increase functionality
  - Event emission verification
  - Edge cases and error conditions

## Related PRs and Issues

Closes #77
Closes #78
Closes #79 
Closes #80

### CODE REVIEW

File: contracts/core/SedaCoreV1.sol [hunk 1]
```diff
@@ -65,11 +79,32 @@ contract SedaCoreV1 is ISedaCore, RequestHandlerBase, ResultHandlerBase, UUPSUpg
...
+        // Check that amount is equal to sum of fees
+        if (msg.value != requestFee + resultFee + batchFee) {
+            revert InvalidFeeAmount();
+        }
+
         bytes32 requestId = super.postRequest(inputs);
         _addRequest(requestId);
-        // Store the request timestamp
-        _storageV1().requestTimestamps[requestId] = block.timestamp;
+        // Store the request details
```
Comment (gluax) [inline contracts/core/SedaCoreV1.sol:99]: Example: it's quite obvious we are storing the request details, so why are we doing it would be a better comment.

File: contracts/core/SedaCoreV1.sol [hunk 2]
```diff
@@ -65,11 +79,32 @@ contract SedaCoreV1 is ISedaCore, RequestHandlerBase, ResultHandlerBase, UUPSUpg
...
+        return postRequest(inputs, 0, 0, 0);
+    }
+
+    function postRequest(
+        SedaDataTypes.RequestInputs calldata inputs,
+        uint256 requestFee,
+        uint256 resultFee,
+        uint256 batchFee
+    ) public payable override returns (bytes32) {
+        // Check that amount is equal to sum of fees
```
Comment (gluax) [inline contracts/core/SedaCoreV1.sol:92]: Another example: of where we should do a _why_ not a _what_.

File: contracts/core/SedaCoreV1.sol [hunk 3]
```diff
@@ -80,22 +115,101 @@ contract SedaCoreV1 is ISedaCore, RequestHandlerBase, ResultHandlerBase, UUPSUpg
...
+                // If no payback address, send all request fee to requestor
+                refundAmount += requestDetails.requestFee;
+            } else {
+                // Calculate request fee split and send to payback address
+                uint256 submitterFee = (result.gasUsed * requestDetails.requestFee) / requestDetails.gasLimit;
+                if (submitterFee > 0) {
+                    _transferFee(payableAddress, submitterFee);
+                    emit FeeDistributed(result.drId, payableAddress, submitterFee, ISedaCore.FeeType.REQUEST);
+                }
+                // Update requestor amount with refund (executed at the end of the function)
```
Comment (gluax) [inline contracts/core/SedaCoreV1.sol:153]: For example, here we could instead say:
```
//  We need to calculate the refund any remaining fees to pay back to the requestor, however,
// we can't pay out the amount till correctness is ensured.
```

It's more verbose, but imo it's significantly more clear to even a new person reading the code.

File: contracts/core/SedaCoreV1.sol [hunk 4]
```diff
@@ -80,22 +115,101 @@ contract SedaCoreV1 is ISedaCore, RequestHandlerBase, ResultHandlerBase, UUPSUpg
...
     }
 
+    /// @inheritdoc ISedaCore
+    function increaseFees(
+        bytes32 requestId,
+        uint256 additionalRequestFee,
+        uint256 additionalResultFee,
+        uint256 additionalBatchFee
+    ) public payable override {
+        // Check that amount is equal to sum of fees
```
Comment (gluax) [inline contracts/core/SedaCoreV1.sol:192]: One more example and I'll stop yapping about it

`// We have to ensure the total fees match the expected amount to prevent incorrect payments or underfunding.`

**Review by gluax — APPROVED** [review]
The code logic looks good to me.

I'm going to nit on the comments as the auditors want comment coverage.
Some of the comments describe _**what**_ the code is doing. Which should only be done if something complex is happening, which I don't think most of this is.
Comments should describe _**why**_ the code is doing what it's doing. This helps maintain readability and understanding more than stating what something is doing.

The comments on functions and types are good though :).

File: contracts/core/SedaCoreV1.sol [hunk 5]
```diff
@@ -65,11 +79,32 @@ contract SedaCoreV1 is ISedaCore, RequestHandlerBase, ResultHandlerBase, UUPSUpg
...
-        // Store the request timestamp
-        _storageV1().requestTimestamps[requestId] = block.timestamp;
+        // Store the request details
+        _storageV1().requestDetails[requestId] = RequestDetails({
+            requestor: msg.sender,
+            timestamp: block.timestamp,
+            requestFee: requestFee,
+            resultFee: resultFee,
+            batchFee: batchFee,
+            gasLimit: inputs.execGasLimit + inputs.tallyGasLimit
```
[the remark this answers is not in the document]
Comment (mariocao) [reply contracts/core/SedaCoreV1.sol:105]: Should be okay because DR id is derived also for those fields.
The only reason for this redundancy is avoid one extra read (fetching the data request by id) when posting a result and calculating the fee split.

File: contracts/core/SedaCoreV1.sol [hunk 6]
```diff
@@ -65,11 +75,35 @@ contract SedaCoreV1 is ISedaCore, RequestHandlerBase, ResultHandlerBase, UUPSUpg
...
         SedaDataTypes.RequestInputs calldata inputs
-    ) public override(RequestHandlerBase, IRequestHandler) returns (bytes32) {
-        bytes32 requestId = super.postRequest(inputs);
+    ) public payable override(RequestHandlerBase, IRequestHandler) returns (bytes32) {
+        return postRequest(inputs, 0, 0, 0);
+    }
+
+    function postRequest(
+        SedaDataTypes.RequestInputs calldata inputs,
+        uint256 requestFee,
```
Comment (FranklinWaller) [inline contracts/core/SedaCoreV1.sol:84]: Nothing to fix here but we should really have a wrapper or something to simplify the 3 fees and the rest of the inputs properties. (even if the user is overpaying, they'll get it refunded.)

File: contracts/mocks/SedaPermissioned.sol [hunk 7]
```diff
@@ -135,6 +159,12 @@ contract SedaPermissioned is ISedaCore, RequestHandlerBase, AccessControl, Pausa
         return queriedPendingRequests;
     }
 
+    /// @inheritdoc ISedaCore
+    /// @dev This is a mock implementation that does nothing and requires admin role
```
Comment (Thomasvdam) [inline contracts/mocks/SedaPermissioned.sol:163]: It doesn't require admin right?

**Review by FranklinWaller — APPROVED** [review]

**Review by Thomasvdam — APPROVED** [review]

File: contracts/mocks/SedaPermissioned.sol (hunk 7 above)
Comment (mariocao) [reply contracts/mocks/SedaPermissioned.sol:163]: Tbh, I was tempted to remove the SedaPermissioned contract

Comment (Thomasvdam) [reply contracts/mocks/SedaPermissioned.sol:163]: Yeah I missed that it was the permissioned/mock version when I made the comment :p
